homeworks/solutions/vseup/2-ddg/src/halfedge.js

```diff
@@ -0,0 +1,529 @@
+    let vertexB = vertices[1].position;
+    let vertexC = vertices[2].position;
+
+    let lengthBC = vertexB.sub(vertexC).norm();
+    let lengthCA = vertexC.sub(vertexA).norm();
+    let lengthAB = vertexA.sub(vertexB).norm();
+
+    let barry1 = lengthBC * lengthBC * (lengthCA * lengthCA + lengthAB * lengthAB - lengthBC * lengthBC);
+    let barry2 = lengthCA * lengthCA * (lengthAB * lengthAB + lengthBC * lengthBC - lengthCA * lengthCA);
+    let barry3 = lengthAB * lengthAB * (lengthBC * lengthBC + lengthCA * lengthCA - lengthAB * lengthAB);
+
+    let barycenter = new Vector(barry1, barry2, barry3);
+
+    return barycenter;
+
+  }
+  /**
+   * returns all 3 vertices of the face
+   */
+  getVertices() {
+    let halfedges = [this.halfedge, this.halfedge.next, this.halfedge.prev];
+    let vertexA = halfedges[0].vertex;
+    let vertexB = halfedges[1].vertex;
+    let vertexC = halfedges[2].vertex;
+    return [vertexA, vertexB, vertexC];
+  }
+}
+
+class Vertex {
+  constructor() {
+    this.position = null // Vector
+    this.halfedge = null // Halfedge
+    this.idx = -1   // Number
+  }
+  normal(method = 'equal-weighted') {
+
+    let incidentFaces = this.getIncidentFaces();
+    let normalSum = new Vector();
+
+    switch (method) {
+      case 'equal-weighted':
+
+        // TODO: compute euqally weighted normal of this vertex
+        for (let face of incidentFaces) {
+          normalSum = normalSum.add(face.getNormal());
+        }
+        return normalSum.unit();
+
+      case 'area-weighted':
+
+        // TODO: compute area weighted normal of this vertex
+
+        for (let face of incidentFaces) {
+          let area = face.getArea();
+          normalSum = normalSum.add(face.getNormal().scale(area));
+        }
+        return normalSum.unit();
+
+      case 'angle-weighted':
+        // TODO: compute angle weighted normal of this vertex
+
+        for (let face of incidentFaces) {
+          let angle = face.getAngle(this.idx);
+          normalSum = normalSum.add(face.getNormal().scale(angle));
+        }
+        return normalSum.unit();
+
+
+      default: // undefined
+        return new Vector()
+    }
+  }
+  curvature(method = 'Mean') {
+
+    let H = this.getMeanCurvature();
+    let K = this.getGaussianCurvature();
+
+    switch (method) {
+      case 'Mean':
+        // TODO: compute mean curvature
+        return K < 0 ? -H : H
+
+      case 'Gaussian':
+        // TODO: compute Guassian curvature
+        return K;
+
+      case 'Kmin':
+        // TODO: compute principal curvature and return Kmin
+        return H - Math.sqrt(Math.abs(H * H - K));
+
+      case 'Kmax':
+        // TODO: compute principal curvature and return Kmax
+        return H + Math.sqrt(Math.abs(H * H - K));
+
+      default: // undefined
+        return 0
+    }
+  }
+
+  // NOTE: you can add more methods if you need here
+  /**
+   * returns all faces incident to the vertex
+   */
+  getIncidentFaces() {
+    let incidentFaces = [];
```

Comment:
This actually causes a lot of allocation for each of your vertices. For our project purpose, it is good as it is, when you start dealing with large meshes, you might want to switch to iterator pattern (see reference solution) for implementing this method, and the caller may stop in the middle of traversal (when we don't need access all faces). Similar to the other traversal methods.